courseworks/week5/given_trisym.jl

```diff
@@ -0,0 +1,72 @@
+using LinearAlgebra,Test, Random
+
+#following code taken from notebook
+struct GivensMatrix{T} <: AbstractArray{T, 2}
+	c::T
+	s::T
+	i::Int
+	j::Int
+	n::Int
+end
+
+Base.size(g::GivensMatrix) = (g.n, g.n)
+
+Base.size(g::GivensMatrix, i::Int) = i == 1 || i == 2 ? g.n : 1
+
+function givens(A, i, j)
+	x, y = A[i, 1], A[j, 1]
+	norm = sqrt(x^2 + y^2)
+	c = x/norm
+	s = y/norm
+	return GivensMatrix(c, s, i, j, size(A, 1))
+end
+
+function left_mul!(A::AbstractMatrix, givens::GivensMatrix)
+	for col in 1:size(A, 2)
+		vi, vj = A[givens.i, col], A[givens.j, col]
+		A[givens.i, col] = vi * givens.c + vj * givens.s
+		A[givens.j, col] = -vi * givens.s + vj * givens.c
+	end
+	return A
+end
+function right_mul!(A::AbstractMatrix, givens::GivensMatrix)
+	for row in 1:size(A, 1)
+		vi, vj = A[row, givens.i], A[row, givens.j]
+		A[row, givens.i] = vi * givens.c + vj * givens.s
+		A[row, givens.j] = -vi * givens.s + vj * givens.c
+	end
+	return A
+end
+
```

Comment:
here for our case we may not need to compute the result for all row/col in 1:size(A, 1 or 2) because we are considering a tridiagonal matrix.
and @inbounds can be added before the loops

Reply:
Thanks, I changed related code!